azurerm/resource_arm_eventgrid_domain.go

```diff
@@ -0,0 +1,380 @@
+	ctx := meta.(*ArmClient).StopContext
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resGroup := id.ResourceGroup
+	name := id.Path["domains"]
+
+	future, err := client.Delete(ctx, resGroup, name)
+	if err != nil {
+		if response.WasNotFound(future.Response()) {
+			return nil
+		}
+		return fmt.Errorf("Error deleting Event Grid Domain %q: %+v", name, err)
+	}
+
+	if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		if response.WasNotFound(future.Response()) {
+			return nil
+		}
+		return fmt.Errorf("Error deleting Event Grid Domain %q: %+v", name, err)
+	}
+
+	return nil
+}
+
+func expandAzureRmEventgridDomainInputMapping(d *schema.ResourceData) *eventgrid.JSONInputSchemaMapping {
+	imf, imfok := d.GetOk("input_mapping_fields")
+
+	imdv, imdvok := d.GetOk("input_mapping_default_values")
+
+	if !imfok && !imdvok {
+		return nil
+	}
+
+	jismp := eventgrid.JSONInputSchemaMappingProperties{}
+
+	if imfok {
+		mappings := imf.([]interface{})
+		mapping := mappings[0].(map[string]interface{})
+
+		if id := mapping["id"].(string); id != "" {
+			jismp.ID = &eventgrid.JSONField{SourceField: &id}
+		}
+
+		if eventTime := mapping["event_time"].(string); eventTime != "" {
+			jismp.EventTime = &eventgrid.JSONField{SourceField: &eventTime}
+		}
+
+		if topic := mapping["topic"].(string); topic != "" {
+			jismp.Topic = &eventgrid.JSONField{SourceField: &topic}
+		}
+
+		if dataVersion := mapping["data_version"].(string); dataVersion != "" {
+			jismp.DataVersion = &eventgrid.JSONFieldWithDefault{SourceField: &dataVersion}
+		}
+
+		if subject := mapping["subject"].(string); subject != "" {
+			jismp.Subject = &eventgrid.JSONFieldWithDefault{SourceField: &subject}
+		}
+
+		if eventType := mapping["event_type"].(string); eventType != "" {
+			jismp.EventType = &eventgrid.JSONFieldWithDefault{SourceField: &eventType}
+		}
+	}
+
+	if imdvok {
+		mappings := imdv.([]interface{})
+		mapping := mappings[0].(map[string]interface{})
+
+		if dataVersion := mapping["data_version"].(string); dataVersion != "" {
+			jismp.DataVersion = &eventgrid.JSONFieldWithDefault{DefaultValue: &dataVersion}
+		}
+
+		if subject := mapping["subject"].(string); subject != "" {
+			jismp.Subject = &eventgrid.JSONFieldWithDefault{DefaultValue: &subject}
+		}
+
+		if eventType := mapping["event_type"].(string); eventType != "" {
+			jismp.EventType = &eventgrid.JSONFieldWithDefault{DefaultValue: &eventType}
+		}
+	}
+
+	jsonMapping := eventgrid.JSONInputSchemaMapping{
+		JSONInputSchemaMappingProperties: &jismp,
+		InputSchemaMappingType:           eventgrid.InputSchemaMappingTypeJSON,
+	}
+
+	return &jsonMapping
+}
+
+func flattenAzureRmEventgridDomainInputMapping(input eventgrid.BasicInputSchemaMapping) []interface{} {
+	if input == nil {
+		return nil
+	}
+	result := make(map[string]interface{})
+
+	jsonValues := input.(eventgrid.JSONInputSchemaMapping).JSONInputSchemaMappingProperties
```

Comment:
whilst it may not be possible in the current SDK version, if an additional type is introduced it could be possible for this cast to fail; as such can we catch this:

```
jsonValues, ok := input.(eventgrid.JSONInputSchemaMapping)
if !ok {
  return nil, fmt.Errorf("")
}

props := jsonValues.JSONInputSchemaMappingProperties
```